File: app/src/main/java/org/digitalcampus/oppia/adapter/RecyclerViewClickableAdapter.java
package org.digitalcampus.oppia.adapter;

import android.view.View;

import androidx.recyclerview.widget.RecyclerView;

public abstract class RecyclerViewClickableAdapter<H extends RecyclerViewClickableAdapter.ViewHolder> extends RecyclerView.Adapter<H> {

    public interface OnItemClickListener {
        void onItemClick(View v, int position);
    }

    private OnItemClickListener itemClickListener;

    public void setOnItemClickListener(OnItemClickListener listener) {
        this.itemClickListener = listener;
    }

    public abstract class ViewHolder extends RecyclerView.ViewHolder {

        public ViewHolder(View itemView) {
            super(itemView);

            itemView.setOnClickListener(v -> {
                if (itemClickListener != null) {
                    itemClickListener.onItemClick(v, getAdapterPosition());
                }
            });
        }
    }

}

File: app/src/main/java/org/digitalcampus/oppia/adapter/RecyclerViewClickableAdapter.kt
package org.digitalcampus.oppia.adapter

import android.view.View
import androidx.recyclerview.widget.RecyclerView
import org.digitalcampus.oppia.listener.OnItemClickListener

abstract class RecyclerViewClickableAdapter<H : RecyclerViewClickableAdapter<H>.ViewHolder> :
    RecyclerView.Adapter<H>() {

    private var itemClickListener: OnItemClickListener? = null

    abstract inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        init {
            itemView.setOnClickListener {
                itemClickListener?.onItemClick(itemView, adapterPosition)
            }
        }
    }

    fun setOnItemClickListener(listener: OnItemClickListener) {
        itemClickListener = listener
    }
}